msg_queue.go

```diff
@@ -35,69 +53,161 @@ func (m *msgQueue) readMessage(st State, current *View) *MessageReq {
-	return &msgQueue{
-		roundChangeStateQueue: msgQueueImpl{},
-		acceptStateQueue:      msgQueueImpl{},
-		validateStateQueue:    msgQueueImpl{},
+func newMsgQueue(logger Logger) *msgQueue {
+	m := &msgQueue{
+		roundChangeStateQueue: &msgQueueImpl{},
+		acceptStateQueue:      &msgQueueImpl{},
+		validateStateQueue:    &msgQueueImpl{},
+		notifyMessageCh:       make(chan struct{}, 1), //hack: there is a bug when you have several messages pushed on the same time.
+	}
+	m.initCommitValidationRoutine(logger)
+	return m
+}
+
+// initCommitValidationRoutine initializes and starts commit validation routine
+// which validates commit messages
+func (m *msgQueue) initCommitValidationRoutine(logger Logger) {
+	m.commitValidation = newCommitValidationRoutine(logger, m.pushMessage)
+	go m.commitValidation.run(nil)
+}
+
+type stateInfo struct {
+	proposalHash []byte
+	validators   ValidatorSet
+	view         *View
+}
+
+// updateStateInfo receives data needed for commit message validation
+// and triggers reading of commit messages
+func (m *msgQueue) updateStateInfo(info *stateInfo) {
+	// feed commit validation routine with necessary data
+	m.commitValidation.updateStateInfoCh <- info
+}
+
+const (
+	maxWorkersCount          = 5
+	pendingCommitsBufferSize = 10
+)
+
+// commitValidationRoutine encapsulates concurrent commit messages validation logic.
+// Valid commit messages are pushed to the validateStateQueue for further consumption by consensus algorithm.
+// Invalid commit messages are discarded.
+type commitValidationRoutine struct {
+	pendingCommitMsgs     *msgQueueImpl
+	lock                  sync.RWMutex
+	logger                Logger
+	validMsgHandler       func(msg *MessageReq)
+	updateStateInfoCh     chan *stateInfo
+	notifyPendingCommitCh chan struct{}
+	closeCh               chan struct{}
+}
+
+// newCommitValidationRoutine creates a new instance of commitValidationRoutine object
+func newCommitValidationRoutine(logger Logger, validMsgHandler func(msg *MessageReq)) *commitValidationRoutine {
+	return &commitValidationRoutine{
+		pendingCommitMsgs:     &msgQueueImpl{},
+		validMsgHandler:       validMsgHandler,
+		logger:                logger,
+		updateStateInfoCh:     make(chan *stateInfo),
+		notifyPendingCommitCh: make(chan struct{}),
+		closeCh:               make(chan struct{}),
+	}
+}
+
+// run contains main part of the commit messages validation logic
+func (c *commitValidationRoutine) run(doneCh chan struct{}) {
+	if doneCh == nil {
+		doneCh = make(chan struct{})
+	}
+	// validateMsgWorker validates single commit message
+	validateMsgWorker := func(commitMsgsCh <-chan *MessageReq, stateInfo *stateInfo) {
+		for commitMsg := range commitMsgsCh {
+			if err := stateInfo.validators.VerifySeal(commitMsg.From, commitMsg.Seal, stateInfo.proposalHash); err != nil {
+				doneCh <- struct{}{}
+				c.logger.Printf("[ERROR] Commit message is invalid (%v). Error: %s", commitMsg, err)
+				return
+			}
+			doneCh <- struct{}{}
```

Comment:
Purpose of `doneCh` is to block until all messages are processed and I need it because of unit test (https://github.com/0xPolygon/pbft-consensus/pull/80/files#diff-353efbbf5c50447329b1d508abf3f0d44552fd09f810ba1364d67bf5d7ef7f08R165). However even with this change, unit test is flaky for some reason.

Do you have any idea how to figure out when `commitValidationRoutine` is done? Tricky part to me is that we can have more messages than workers and therefore not sure what is the best pattern to figure out when all the messages we have pushed into the queue are processed...? :thinking: